### PR TITLE
Hack test

### PR DESCRIPTION
**Pull request by djarecka**
Removed many general parts from the lesson and expanded regression part.

Haven't work on markdown syntax yet.

### CODE REVIEW

File: _episodes/06-testing.md [hunk 1]
```diff
@@ -23,16 +23,17 @@ keypoints:
 >  - How to run Docker container with CI platform?
 
 
-This lesson extends the software testing subject 
-and introduces Continuous Integration workflow. Examples of Python testing frameworks 
-and CI platforms will be used.   
+This lesson extends the software testing subject and introduces Continuous Integration
```
Comment (satra) [inline _episodes/06-testing.md]: subject?

Comment (satra) [reply _episodes/06-testing.md]: perhaps `topic` is a better word.

File: _episodes/06-testing.md [hunk 2]
```diff
@@ -23,16 +23,17 @@ keypoints:
...
 - Element 1: Unit, integration and regression tests
 - Element 2: Python testing frameworks
-- Element 3: Continuous Integration practice
-- Element 4: Continuous Integration Service: Travis CI
+- Element 3: Regression tests for Simple Workflow
+- Element 4: Continuous Integration practice
+- Element 5: Continuous Integration Service: Travis CI
 
 
 ### Lesson requirements
```
Comment (satra) [inline _episodes/06-testing.md:39]: i would simply point to the software carpentry unit tests here. no need to repeat the basics of unit tests here. part of the goal here is to focus on imaging researchers/developers more than the general audience.

File: _episodes/06-testing.md [hunk 3]
```diff
@@ -47,67 +48,70 @@ Although not essential it is helpful to have an understanding of:
...
+  - you don't have to knows the expected result, the assumption is that the past results were correct.
 
-- [Regression tests](https://en.wikipedia.org/wiki/Regression_testing) verify that software previously developed and tested still 
-  performs correctly even after it was changed or interfaced with other software. 
-  For regression tests you don't have to knows what the expected result should be, 
-  the assumption is that the past results were correct. 
 
+In science, regression tests perform a special role. 
+Scientists often don't know what is the correct answer before doing the research, 
+therefore results can't be compared to the priory known results.
```
Comment (satra) [inline _episodes/06-testing.md]: priory -> prior

File: _episodes/06-testing.md [hunk 4]
```diff
@@ -47,67 +48,70 @@ Although not essential it is helpful to have an understanding of:
...
+However, you still want to be sure that your published results are not sensitive to 
+an operating system, a specific version of library, etc.
+If your results differ when you're changing computer environment, you should understand the 
+source of the changes. 
+The first step would be automatically check the results with various environments
```
Comment (satra) [inline _episodes/06-testing.md]: `be automatically` --> `be to automatically`

File: _episodes/06-testing.md [hunk 5]
```diff
@@ -47,67 +48,70 @@ Although not essential it is helpful to have an understanding of:
...
+whenever you make any changes.
+You can achive this by combining your regression tests with a CI platform.
 
-- Resources: More about unit, integration and regression test you might find at the software carpentry [lesson](http://katyhuff.github.io/python-testing/index.html).
+Regression tests (as well as unit and any other type of tests) can help you with future 
+development process. 
+Whenever you're changing your code, by adding new functionality or improving performance, 
+you can check if you're still able to reproduce your previous results.
```
Comment (satra) [inline _episodes/06-testing.md]: i think this should be clarified into a callout.

changing functionality may change output, which is ok. but it's important to understand what aspect of the code change created a change in output. thus regression testing can help associate changes in output with changes in code.

File: _episodes/06-testing.md [hunk 6]
```diff
@@ -47,67 +48,70 @@ Although not essential it is helpful to have an understanding of:
...
-a python unit test framework, no boilerplate, can run doctests, unittests.  
-  * [pytest library](http://doc.pytest.org/en/latest/): 
-a powerful python test framework, no boilerplate, easy to start working, 
-    covers extensive options and testing features. 
 
-- Resources: a very good introduction to all Python test frameworks by 
-[Brian Okken](http://pythontesting.net/start-here/).
+### Element 2: Testing Python code with Pytest
 
+A very good introduction to all Python test frameworks you can find
```
Comment (satra) [inline _episodes/06-testing.md]: You can find a very good introduction to all Python test frameworks in ...

File: _episodes/06-testing.md [hunk 7]
```diff
@@ -183,40 +172,35 @@ $ `pip install -r requirements.txt`
...
 
-As a developer you should try to commit at least once per day in order to keep track 
-of your changes.
-It is also good to integrate your changes with the main code frequently.
-By doing this, you can find out if there's a conflict between your work 
-and other developers work.
-If the main branch is automatically build and tested, 
-you can easily  detect conflicts in compilation and execution of your code. 
+A common practice is that every single pull request to the main branch of
+the project repository is automatically build and tested before merging.
```
Comment (satra) [inline _episodes/06-testing.md]: built

File: _episodes/06-testing.md [hunk 8]
```diff
@@ -261,43 +245,16 @@ If you're interested, you can find blog posts that compare these tools, e.g by
...
->  - Add new tests to repository and run them using Travis.
->  - You migh want to create a Docker container and use it with the Travis platform.
\ No newline at end of file
+>  - Fork the Simple Workflow repository and clone your own fork to your computer.
+>  - Create a test directory and write a few unit tests for the function from `run_demo_workflow.py`. Run the test locally on your computer. 
+>  - Write some regression tests and compare your results to the results from `expected_output`. Run the test locally.
+>  - Login to Travis and link your repository.
+>  - Create `.travis.yml` file and run your tests using the Travis platform.
+>  - Try to use other CI platforms, e.g. CircleCI to do the same.
+>  - Try to run your tests using a Docker container (see the [README](https://github.com/ReproNim/simple_workflow#1-to-execute-demo-with-docker)) on a CI platform (note that various CI platform have various memory and time limitations).
```
Comment (satra) [inline _episodes/06-testing.md]: all hands on exercises or other exercises should have solutions. also make sure to add exercises as people are going through the lesson, not just at the end.

Comment (djarecka) [reply _episodes/06-testing.md]: @satra - do you think that this exercise works fine here? By solution you mean that I create a repo with tests?

Comment (satra) [reply _episodes/06-testing.md]: Each exercise should have solutions that they can follow. So the exercises have to be a lot more specific with instructions. For example, how do they create the .travis.yml file, etc.,.

Comment (satra) [reply _episodes/06-testing.md]: This is what we have to keep refining with this modules, to really turn them into lessons.

**Pull request merged**